core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -480,6 +480,41 @@ trait EntityComponent {
       }
     }
 
+    private object CopyEntitiesQuery extends RawSqlQuery {
+      val driver: JdbcProfile = EntityComponent.this.driver
+
+      def copyEntities(clonedWorkspaceId: UUID, newWorkspaceId: UUID): WriteAction[Int] = {
+
+        sqlu"""insert into ENTITY (name, entity_type, workspace_id, record_version,
+                    all_attribute_values, deleted, deleted_date)
+                select name, entity_type, $newWorkspaceId, record_version,
+                       all_attribute_values, deleted, deleted_date
+                from ENTITY where workspace_id = $clonedWorkspaceId and deleted = 0
+          """
+      }
+    }
+
+    private object CopyEntityAttributesQuery extends RawSqlQuery {
+      val driver: JdbcProfile = EntityComponent.this.driver
+
+      def copyAllAttributes(clonedWorkspaceId: UUID, newWorkspaceId: UUID): WriteAction[Int] = {
+
+        val sourceShardId = determineShard(clonedWorkspaceId)
+        val destShardId = determineShard(newWorkspaceId)
+        sqlu"""insert into ENTITY_ATTRIBUTE_#$destShardId (name, value_string, value_number, value_boolean, value_entity_ref,
+                list_index, owner_id, list_length, namespace, VALUE_JSON, deleted, deleted_date)
+                SELECT old_entity.name, old_entity.value_string,
+                       old_entity.value_number, old_entity.value_boolean, old_entity.value_entity_ref,
+                       old_entity.list_index, new_entity.id, old_entity.list_length, old_entity.namespace,
+                       old_entity.VALUE_JSON, old_entity.deleted, old_entity.deleted_date
+                from ENTITY new_entity
+                JOIN (select ea.*, e.name as entity_name, e.entity_type from ENTITY_ATTRIBUTE_#$sourceShardId ea join ENTITY e
+                    on ea.owner_id = e.id where e.workspace_id = $clonedWorkspaceId and e.deleted = 0 and ea.deleted = 0) old_entity
+                on new_entity.name = old_entity.entity_name and new_entity.entity_type = old_entity.entity_type
+                where new_entity.workspace_id = $newWorkspaceId;
+          """
```

Comment:
you'll need to do a similar mapping of old-entity-id to new-entity-id inside the attributes' `value_entity_ref` column. Else, any entity references that get copied will still be pointing at the old entities in the source workspace.

Reply:
Good catch! ty

Reply:
Phew this is a thorny query! I know you said you tried a few different approaches and this is the most efficient one. I'll toss out one approach - if you tried it and eliminated it, all good. What about creating a temporary table with two columns: oldEntityId, newEntityId (based on your existing self-join on the entity type&name), then using that temp table to fill in the proper ids for the main INSERT INTO … SELECT statement, both for ENTITY_ATTRIBUTE.owner_id and for  ENTITY_ATTRIBUTE. value_entity_ref ?

Reply:
I've changed the query and I think it's a little bit clearer.  I took your suggestion about creating a dedicated `existing entity id -> new cloned entity id` table and the above is what ultimately fell out of it.  Initially I tried using `create temporary table` as you were suggesting but you [can't reference a temporary table multiple times under different aliases](https://dev.mysql.com/doc/refman/8.0/en/temporary-table-problems.html) and we need to join back to that mapping twice to change both the `owner_id` and `value_entity_ref` so I'd have to create two temporary tables and that all is more verbose than what I ended up with.  

The derived tables `owner_entity` and `referenced_entity` are the same sql, but the same issue arises where I can't join to owner_entity twice for example.  I experimented a bit with trying to extract 
```
select old_entity.id as old_id, new_entity.id as new_id from ENTITY old_entity
                               join ENTITY new_entity on new_entity.entity_type = old_entity.entity_type
                               and new_entity.name = old_entity.name
                               and new_entity.workspace_id = $newWorkspaceId
                               and old_entity.workspace_id = $clonedWorkspaceId
```
to a variable but the parameters in that sql string complicates things.  I think we could probably do something similar to https://github.com/broadinstitute/rawls/blob/e1348ab636351f3f2799c73b6de6b3d0ab40e60a/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala#L229 but I think that approach requires something like `UNHEX(${clonedWorkspaceId.toPrettyString})` so as to get a sql string that no longer has parameters.  That didn't seem like a worthwhile tradeoff to me.  I also tried escaping the $ in the extracted string with $$ but that didn't work for me either. @davidangb if you know of a clean way to make that sql a bit DRYer that might be worthwhile, but I also like the fact that a developer can more quickly see the query being used.

Reply:
To DRY this and store the subquery in a reusable variable, I think you'd need to rely on `concatSqlActions`. There are a bunch of examples of this in the codebase, you'd end up with something like:
```scala
        val entityIdMappingOldToNew = sql"""(select old_entity.id as old_id, new_entity.id as new_id from ENTITY old_entity
                                               join ENTITY new_entity on new_entity.entity_type = old_entity.entity_type
                                               and new_entity.name = old_entity.name
                                               and new_entity.workspace_id = $newWorkspaceId
                                               and old_entity.workspace_id = $clonedWorkspaceId)"""

        concatSqlActions(
          sql"""insert into ENTITY_ATTRIBUTE_#$destShardId (name, value_string, value_number, value_boolean, value_entity_ref,
                list_index, owner_id, list_length, namespace, VALUE_JSON, deleted, deleted_date)
                select ea.name, value_string, value_number, value_boolean, referenced_entity.new_id as value_entity_ref, list_index, owner_entity.new_id as owner_id,
                list_length, namespace, VALUE_JSON, 0, null from ENTITY_ATTRIBUTE_#$sourceShardId ea
                    join """, entityIdMappingOldToNew, sql""" owner_entity
                    on ea.owner_id = owner_entity.old_id
                    left join """, entityIdMappingOldToNew, sql""" referenced_entity
                    on ea.value_entity_ref = referenced_entity.old_id
                join ENTITY e on e.id = ea.owner_id where ea.deleted = false and e.deleted = false and e.workspace_id = $clonedWorkspaceId
          """
        ).as[Int].head
```

which is arguably not any easier to read. Maybe just add code comments explaining the situation? I am fine with either implementation, what you have in place right now looks pretty good and readable.